docs/architecture/adrs/005-chat-completion-support.md

```diff
@@ -0,0 +1,74 @@
+    - We will add a `detectors` block in the request, containing `input` and `output` blocks. This will provide flexibility and control, allowing users to provide list of detectors with parameters for both chat completions input and output separately.
+    - Here, `input` and `output` does not strictly refer to whether only the `input` to chat completions is provided to the detectors or that only the `output` to chat completions is provided to the detectors. Detectors specified in `output` can potentially take both input and output of a chat completions model.
+    - We will add a `detections` block in the response.
+- If no input or output detectors were requested, then the `detections.<missing element / input or output>` would not be present in the response.
+- If none of input or output detectors are requested, then we will return `422` (similar behavior to other standalone `v2` endpoints).
+
+
+### Response behavior
+
+#### Assumptions regarding choices
+This is particular to the [`choices` array](https://platform.openai.com/docs/api-reference/chat/object#chat/object-choices) on the chat completion response.
+1. Each choice is independent.
+2. Each choice is generated with the same input.
+
+#### Response
+- When output detectors are requested, if the LLM (model used to generate chat completions) doesn't generate `content` in any `message` in `choices`, then we will ignore the `choices` and return a warning about `detections.output`.
+- We will run all detectors included in `detectors.output` of the user request on all `choices` (that have `message.content`), unless there are certain detectors that are supposed to be run across choices.
+- We will order all detections that include span by `start` and move all the other ones towards the end (by not ordering them in any particular order), but grouping the remaining ones based on `detector_id`.
+    - An alternate consideration made here was to not order the detections and just include `detector_id`. However, since detectors are called in parallel, just collecting results may lead to different ordering each call. This ordering will allow processing of detections with spans in order by `start`.
+
+
+##### Streaming response
+
+The streaming response with guardrails is built off the [chat completion chunk responses](https://platform.openai.com/docs/api-reference/chat/streaming). This applies when both streaming and output detectors are requested.
+- Each choice's `delta.content`, if present, will be the `content` used for `detections`. Like the unary case, if there is no `content` overall, a warning will be returned.
+- Since detector use will cause content to be "chunked" (e.g. collected into sentences for sentence detectors), to prevent ambiguity, the `delta.role` will be provided on each stream event, and `delta.content` will include the "chunked" text. This will apply to each `choice` in `choices`.
```

Comment:
Can we add a point here that this API will also follow the "only returning fully processed chunks" strategy, same as text generation ?

https://github.com/foundation-model-stack/fms-guardrails-orchestrator/blob/main/docs/architecture/adrs/002-streaming-response-aggregation.md#:~:text=Only%20fully%20processed%20frames%20will